code/game/machinery/keypads.dm

```diff
@@ -0,0 +1,237 @@
+				return
+			W.loc = src
+			device = W
+			user << "<span class='notice'>You add [W] to \the [src].</span>"
+
+		if(wired && istype(W, /obj/item/weapon/wirecutters))
+			var/obj/item/stack/cable_coil/C = new /obj/item/stack/cable_coil(loc)
+			C.amount = 5
+			user << "<span class='notice'>You cut the wires.</span>"
+			stat |= MAINT
+			wired = 0
+			update_icon()
+
+		if(!device && !wired && istype(W, /obj/item/weapon/wrench))
+			user << "<span class='notice'>You start unsecuring \the [src] frame...</span>"
+			playsound(loc, 'sound/items/Ratchet.ogg', 50, 1)
+			if(do_after(user, 40/W.toolspeed, target = src))
+				user << "<span class='notice'>You unsecure \the [src] frame.</span>"
+				var/obj/item/wallframe/keypad/K = new(get_turf(src))
+				transfer_fingerprints_to(K)
+				K.password = password
+				playsound(loc, 'sound/items/Deconstruct.ogg', 50, 1)
+				qdel(src)
+
+		update_icon()
+		return
+
+	return attack_hand(user)
+
+/obj/machinery/keypad/emag_act(mob/user)
+	user << "<span class='warning'>The password has been set to nothing!</span>"
+	password = ""
+	playsound(loc, "sparks", 100, 1)
+
+/obj/machinery/keypad/attack_ai(mob/user)
+	if(!panel_open)
+		return attack_hand(user)
+
+/obj/machinery/keypad/attack_hand(mob/user)
+	add_fingerprint(user)
+	if(panel_open)
+		if(device)
+			if(device)
+				device.loc = get_turf(src)
+				device = null
+			update_icon()
+			user << "<span class='notice'>You remove electronics from \the [src] frame.</span>"
+		return
+
+	if((stat & (NOPOWER|BROKEN|MAINT)))
+		return
+
+	if(device && device.cooldown)
+		return
+
+	var/dat
+	dat += "<h3>Code:</h3>"
+	if(output != "")
+		dat += output
+	else
+		dat += " "
+		for(var/i=1, i <= length(input), i++)
+			dat += "*"
+	dat += "<h3>Input:</h3>"
+	dat += "<A href='?src=\ref[src];num=1'>1</A><A href='?src=\ref[src];num=2'>2</A><A href='?src=\ref[src];num=3'>3</A><BR>"
+	dat += "<A href='?src=\ref[src];num=4'>4</A><A href='?src=\ref[src];num=5'>5</A><A href='?src=\ref[src];num=6'>6</A><BR>"
+	dat += "<A href='?src=\ref[src];num=7'>7</A><A href='?src=\ref[src];num=8'>8</A><A href='?src=\ref[src];num=9'>9</A><BR>"
+	dat += "<A href='?src=\ref[src];num=C'>C</A><A href='?src=\ref[src];num=0'>0</A><A href='?src=\ref[src];num=E'>E</A><BR>"
+	dat += "</div>"
+
+	var/datum/browser/popup = new(user, "keypad", "Keypad", 240, 240)	//Set up the popup browser window
+	popup.set_title_image(user.browse_rsc_icon(icon, icon_state))
+	popup.set_content(dat)
+	popup.open()
+
+/obj/machinery/keypad/Topic(href, href_list)
+	if(..())
+		return 0
+
+	usr.set_machine(src)
+	if(href_list["num"])
+		if(href_list["num"] == "C")
+			input = ""
+		else if(href_list["num"] == "E")
+			if(input != password)
+				output = "ERROR"
+				flick("[skin]-denied", src)
+			else
+				input = ""
+				use_power(2)
+				flick("[skin]1", src)
+
+				if(device)
+					device.pulsed()
+		else
+			if(length(input) < max_digits)
+				input = input+"[href_list["num"]]"
```

Comment:
always assume that topic can get malicious arguments, this doesn't look very safe,is it